src/communication/NetworkClient.ts

```diff
@@ -92,71 +107,67 @@ export default class NetworkClient {
+    const headers: Record<string, string> = {};
+    headers['User-Agent'] = composeUserAgent(nodeVersion, libraryVersion, versionStrings);
     if (apiKey != undefined) {
-      axiosOptions.headers['Authorization'] = `Bearer ${apiKey}`;
+      headers['Authorization'] = `Bearer ${apiKey}`;
     } /* if (accessToken != undefined) */ else {
-      axiosOptions.headers['Authorization'] = `Bearer ${accessToken}`;
-      axiosOptions.headers['User-Agent'] += ' OAuth/2.0';
+      headers['Authorization'] = `Bearer ${accessToken}`;
+      headers['User-Agent'] += ' OAuth/2.0';
     }
-    axiosOptions.headers['Accept'] = 'application/hal+json';
-    axiosOptions.headers['Accept-Encoding'] = 'gzip';
-    axiosOptions.headers['Content-Type'] = 'application/json';
-    // Create the Axios instance.
-    this.axiosInstance = axios.create({
-      ...axiosOptions,
-      baseURL: apiEndpoint,
-      httpsAgent: new https.Agent({
-        ca: caCertificates,
-      }),
-    });
-    // Make the Axios instance request multiple times is some scenarios.
-    makeRetrying(this.axiosInstance);
+    headers['Accept'] = 'application/hal+json';
+    headers['Accept-Encoding'] = 'gzip';
+    headers['Content-Type'] = 'application/json';
+
+    // Create the https agent.
+    const agent = new https.Agent({ ca: caCertificates });
+
+    // Create retrying fetch function.
+    const fetchWithRetries = retryingFetch(fetch);
+
+    // Create the request function.
+    this.request = (pathname, options) => {
+      const url = new URL(pathname, apiEndpoint);
+      return fetchWithRetries(url, { agent, ...options, headers: { ...headers, ...options?.headers } });
+    };
   }
 
   async post<R>(pathname: string, data: Data & IdempotencyParameter, query?: SearchParameters): Promise<R | true> {
     // Take the idempotency key from the data, if any. It would be cleaner from a design perspective to have the
     // idempotency key in a separate argument instead of cramming it into the data like this. However, having a
     // separate argument would require every endpoint to split the input into those two arguments and thus cause a lot
     // of boiler-plate code.
-    let config: AxiosRequestConfig | undefined = undefined;
-    if (data.idempotencyKey != undefined) {
-      const { idempotencyKey, ...rest } = data;
-      config = { headers: { [idempotencyHeaderName]: idempotencyKey } };
-      data = rest;
-    }
-    const response = await this.axiosInstance.post(buildUrl(pathname, query), data, config).catch(throwApiError);
-    if (response.status == 204) {
-      return true;
-    }
-    return response.data;
+    const { idempotencyKey, ...body } = data;
+    const config: RequestInit = {
+      method: 'POST',
+      headers: idempotencyKey ? { [idempotencyHeaderName]: idempotencyKey } : undefined,
+      body: JSON.stringify(body),
+    };
+    return this.request(buildUrl(pathname, query), config).catch(throwApiError).then(processFetchResponse);
```

Comment:
The `.catch(throwApiError).then(processFetchResponse)` seems repetitive. Can we change the abstraction somehow?